torch/backends/_nnapi/serializer.py

```diff
@@ -0,0 +1,1131 @@
+
+        # NNAPI uses 4 values for padding.
+        ph, pw = paddings
+        real_paddings = [ph, ph, pw, pw]
+
+        return ConvPoolArgs2d(*(kernels + strides + real_paddings + dilations + [group_num]))
+
+
+
+    def serialize_model(self, model, inputs):
+        self.add_immediate_bool_scalar(False)
+        self.add_immediate_bool_scalar(True)
+
+        inp_dim_orders = []
+        out_dim_orders = []
+
+        self_jitval = next(model.graph.inputs())
+        self.add_constant_value(self_jitval, self_jitval.type(), model)
+
+        for input_value, input_tensor in zip(list(model.graph.inputs())[1:], inputs):
+            op_id = self.add_tensor_operand_for_input(input_value, input_tensor)
+            inp_dim_orders.append(self.operands[op_id].dim_order.value)
+
+        for idx, node in enumerate(model.graph.nodes()):
+            LOG.debug("Processing node #%d: %r", idx, node)
+            self.add_node(node)
+
+        retn = model.graph.return_node()
+        assert retn.inputsSize() == 1
+        assert retn.outputsSize() == 0
+        # TODO: Make outputs a local variable? 
+        # TODO: Handle tuple-of-tensor return
+        for idx in range(1):
+            op_id = self.jitval_operand_map[retn.inputsAt(0)]
+            self.outputs.append(op_id)
+            out_dim_orders.append(self.operands[op_id].dim_order.value)
+
+        model = []
+
+        version = 1
+        header = struct.pack(
+                "iiiiii",
+                version,
+                len(self.operands),
+                len(self.values),
+                len(self.operations),
+                len(self.inputs),
+                len(self.outputs),
+                )
+        model.append(header)
+
+        serialized_values, serialized_value_data = self.serialize_values()
+
+        model.extend(struct.pack("iifi", t, len(d), s, z) for (t,d,_m,s,z) in self.operands)
+        model.extend(serialized_values)
+        model.extend(struct.pack("iii", *x) for x in self.operations)
+        model.extend(self.serialize_ints(fix_shape(dims, mf)) for (_, dims, mf, _, _) in self.operands)
+        model.extend(serialized_value_data)
+        model.append(self.serialize_ints(self.operation_args))
+        model.append(self.serialize_ints(self.inputs))
+        model.append(self.serialize_ints(self.outputs))
+
+        #return (b"".join(model), self.used_weight_tensor_names)
+        return (b"".join(model), self.used_weights, inp_dim_orders, out_dim_orders)
+
+
+    def serialize_values(self):
+        serialized_values = []
+        serialized_value_data = []
+        assert len(self.values) == len(self.value_data)
+        for ((op_index, source_type), data) in zip(self.values, self.value_data):
+            source_length = len(data)
+
+            # Pad with 0 bytes out to a multiple of 4 for alignment.
+            physical_length = ((source_length - 1) | 0x3) + 1
+            padded_data = data + (b"\0" * (physical_length - source_length))
```

Comment:
@dreiss Q. Does this assume a certain endianness?

Reply:
I don't think so.  I assume little-endian somewhere else, though.